Guide and Template for writing a Price ID UMIP (#123)
UMAprotocol/UMIPs: UMIPs/Guide-to-writing-a-price-identifier-UMIP.md

> # Guide to writing a UMIP for Price Identifiers

Reviewer: suggest "an UMIP" rather than "a UMIP" and consistent use of indefinate article throughout.

Author: I'm no english major but since UMIP stands for UMA Improvement Proposal (making it a noun or something like that) it seems and sounds weird to have it be an UMIP. I've said it aloud a few times and it doesn't make sense...lmk if you still feel otherwise. Maybe this is one of those scenarios where the 'an' rule does not apply.

## HEADERS
| UMIP [#]     |                                                                                                                                          |
|------------|------------------------------------------------------------------------------------------------------------------------------------------|
| UMIP Title | [TITLE]                                                                                                  |
| Authors    | [Name or username and email]
| Status     | [Draft, Last Call, Approved, Final, Abandoned, Rejected]                                                                                                                             |
| Created    | [DATE]                                                                             
<br>

Note that an UMIP number will be assigned by an editor. When opening a pull request to submit your UMIP, please use an abbreviated title in the filename, umip-draft_title_abbrev.md.
The title should be 44 characters or less.

# Status
A successful UMIP will move along the following stages: Draft ⟶ Last Call ⟶ Final ⟶ Approved. Unsuccessful states are also possible: Abandoned and Rejected.

## Draft

> A UMIP that is open for consideration and is undergoing rapid iteration and changes should be set to “Draft”. In order to proceed to “Last Call", the UMIP template for adding a new price identifier must be completed. Every UMIP author is responsible for facilitating conversations and building community consensus for the proposal in [UMA’s Discord](https://discord.com/invite/jsb9XQJ)  via the “price-id-umips’ channel. 

Reviewer: Suggest that these conversations and consensus building activities are moved to Discourse (once operational).

Maintain the "Price-id-umips" channel in Discord to catch any additional substantive points, but encourage the community to shift substantive convos to Discourse, both through general nudging and targeted nudging if a significant contribution is made in Discord, with admin duplicating if necessary.

## Last Call

> A UMIP that is done with its initial iteration and ready for review by a wide audience will be set to “Last Call” status. If upon review, there is a material change or substantial unaddressed complaints, the UMIP will revert to "Draft". If the UMIP requires code changes, the core devs must review the UMIP. A successful UMIP will be in "Last Call" status for a reasonable period of time for the community to provide feedback before moving to a vote by UMA tokenholders.

Reviewer: Admin to note in the Discourse when a PI UMIP moves from Draft-> Last Call.

## Final

> An UMIP that successfully passes "Last Call" will move to "Final" status and be voted on by UMA tokenholders.

Reviewer: Admin/UMIP proposer to note in the Discourse 
 - that a UMIP is now ready for voting
 - summerise any salient points from the discussions that voters may wish to take note of
 - highlight the date/times that the proposal is open for voting + general reminder about how to vote.

## Approved

If UMA tokenholders approve the proposal, UMA protocol will be updated to include the necessary changes. At this time, any code changes (if relevant) should be merged into the core protocol repository so that the core protocol repository always reflects the live code that is running. The UMIP is then considered to be in the "Approved" state.

## Abandoned

If at any point during the UMIP Finalization Process, a UMIP is abandoned, it will be labeled as such. Grounds for abandonment include a lack of interest by the original author(s), or it may not be a preferred option anymore.

## Rejected
If UMA tokenholders do not approve a proposal, or the UMIP is fundamentally broken or rejected by the community, it will not be implemented.

> <br>

Reviewer: Admin to note in the Discourse what the final outcome of the UMIP vote is, and close the topic.

# SUMMARY 

"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the issue.

**Example**

The DVM should support price requests for the [Base currency / Quote currency] price index.

> 

Reviewer 2: test

# Guide to writing a UMIP for Price Identifiers

Below is a detailed guide for creating a UMIP to register a new price identifier with UMA’s Data Verification Mechanism (DVM). 

## HEADERS

| UMIP [#]     |                                                                                                                                          |
|------------|------------------------------------------------------------------------------------------------------------------------------------------|
| UMIP Title | [TITLE]                                                                                                  |
| Authors    | [Name or username and email]
| Status     | [Draft, Last Call, Approved, Final, Abandoned, Rejected]                                                                                                                                   |
| Created    | [DATE]                                                                             
<br>

> Note that an UMIP number will be assigned by an editor. When opening a pull request to submit your UMIP, please use an abbreviated title in the filename, umip-draft_title_abbrev.md.

Reviewer 3: I think we should move this up as ##Contribution Guidelines above ##Headers

The title should be 44 characters or less.

> # Status

Reviewer 3: In my opinion, this document should be content specific rather than trying to include everything. Because there is a lot! My suggestion would be to remove the status explanations and just reference [umip-1 ](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-1.md) in the `Guide to writing a UMIP for Price Identifiers` section at the top. WDYT?

Reviewer: Suggest that there is a "pre-draft" stage where Price ID UMIP proposers suggest a new PI on Discourse (once operational), as a means to test out community reaction/enthusiasm/concerns
1. To get any initial and obvious issues addressed by the community prior to it a formal draft proposal
2. To act as a "place-setter" for PIs that may be proposed then abandoned so that others may learn from issues/concerns/advantages of previous suggestions of that PI (possible issue that this may become "spammy" tho, unless we can control that section permissions and limit who can propose new PIs, -  may need some thought to the configuration of Discourse).
3. To act as a repository of future discussions as the PI goes through the UMIP process so that discussions are all kept in one place from the start
4. To act as a historical record of discussions around PIs  should later issues be identified.

Reviewer 3: Love the idea of Discourse and fully in support of this. What would be the reason for including another stage though? Could this also be done in the Draft stage?

Author: I like this idea but wouldn't phrase it as a pre-draft phase. I think step 1 is create it in discourse and link to the UMIP on Github. Discourse preserves the converstaion indefinitely whereas others with more nit pick details about the UMIP can address in github. I won't add anything about Discourse until it is live

<br>

 ### **Example** 
<br>

The DVM currently does not support the USD/ETH price index. Supporting the USD/ETH price identifier would enable the creation of zero-coupon fixed-rate dollar loans, if collateralized by WETH. This creates positions similar to using ETH in the MakerDAO system to mint Dai.

Given a current USD/ETH price of .001 (ETH/USD is $1,000), Alice (a minter) would interact with an expiring contract collateralized in WETH by using 0.2 WETH (worth $200) as collateral to mint synthetic tokens at a 200% collateralization ratio. 

When the contract expires at the end of the month, the USDETH price has moved down to .0005 (ETH/USD is $2,000). Alice’s position now has a collateralization ratio of 400% and each synthetic can be redeemed for .0005 WETH. 

<br> 

# MARKETS & DATA SOURCES

### **Required questions**
<br>

1. What markets should the price be queried from? It is recommended to have at least 3 markets. <br> 

    - Which specific pairs should be queried from each market?
2. Provide recommended endpoints to query for real-time prices from each market listed. 
    - These should match the data sources used in your `Price Feed Implementation`. 

>     <br>

Reviewer 3: ```suggestion
```

3. How often is the provided price updated?
4. Provide recommended endpoints to query for historical prices from each market listed. 

    - Do these sources allow for querying up to 74 hours of historical data? 

>     <br>

Reviewer 3: ```suggestion
```

    - How often is the provided price updated?
5. Is an API key required to query these sources?  

    - Is there a cost associated with usage? 

>     <br>

Reviewer 3: ```suggestion
```

    -  If there is a free tier available, how many queries does it allow for?

>      <br>

Reviewer 3: ```suggestion
```

## Rejected
If UMA tokenholders do not approve a proposal, or the UMIP is fundamentally broken or rejected by the community, it will not be implemented.

<br>

# SUMMARY 

"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the issue.

### Example 

The DVM should support price requests for the [Base currency / Quote currency] price index.

<br>

# MOTIVATION

This section should clearly explain the types of financial products that will be created and the mechanics of an example financial product using this price identifier. Please answer the following questions:

1. What are the financial positions enabled by creating this synthetic that do not already exist?
2. Please provide an example of a person interacting with a contract that uses this price identifier. 
    - Remember, price identifiers **return the units of the collateral currency**. 
     - For example, if at expiry the price id returns 1 and the contract is collateralized in renBTC, each synth would be worth 1 renBTC.
3. Consider adding market data  (e.g., if we add a “Dai alternative,” the author could show the market size of Dai)

> <br>

Reviewer 3: ```suggestion
```

When the contract expires at the end of the month, the USDETH price has moved down to .0005 (ETH/USD is $2,000). Alice’s position now has a collateralization ratio of 400% and each synthetic can be redeemed for .0005 WETH. 

<br> 

# MARKETS & DATA SOURCES

### **Required questions**
<br>

1. What markets should the price be queried from? It is recommended to have at least 3 markets. <br> 

    - Which specific pairs should be queried from each market?
2. Provide recommended endpoints to query for real-time prices from each market listed. 
    - These should match the data sources used in your `Price Feed Implementation`. 
    <br>
3. How often is the provided price updated?
4. Provide recommended endpoints to query for historical prices from each market listed. 

    - Do these sources allow for querying up to 74 hours of historical data? 
    <br>

    - How often is the provided price updated?
5. Is an API key required to query these sources?  

    - Is there a cost associated with usage? 
    <br>

    -  If there is a free tier available, how many queries does it allow for?
     <br>

    - What would be the cost of sending 15,000 queries?

<br>

### **Guidance**
<br>

> Currency pairs can have varying prices depending on the market that they are traded on. As an example, the price of BTC/USD will vary between Coinbase Pro, Binance and Gemini. Because of this, you should be specific about the markets that you wish to gather your price information from. 

Reviewer 3: Think we should add in a note for price ids that this might not apply for. Something along the lines of:

```suggestion
Currency pairs can have varying prices depending on the market that they are traded on. As an example, the price of BTC/USD will vary between Coinbase Pro, Binance and Gemini. Because of this, you should be specific about the markets that you wish to gather your price information from. 

For price identifiers that are not determined by market prices, the markets and data sources format may not apply. In these situations, this section should inform users of how they should gather all necessary information used in the price determination process.
```

**This should default to the three highest volume markets available**. If three markets are not available, or you wish to use more or less than three, please specify the justification for doing so.

Specify the availability of this price data. Price data should, at a minimum, meet these requirements.

- Have both a real-time and historical price data sources available
- Data sources should update the price at least once per hour.
- The historical endpoint needs to be able to provide data at least 74 hours in the past
- A liquidation or dispute bot will make approximately 15,000 calls per data source per month. If your recommended data sources are not free, please specify the cost of 15,000 calls.

<br>

> **Note**: An API for accessing data is not the same thing as the markets you choose to determine the price.  You would not say that cryptowatch is a market for data, as it is just an aggregator of market data.  Examples of markets are “Coinbase Pro, Binance, Bitstamp.”  Data sources are still necessary to demonstrate the accessibility of the data and are often just the specific market’s API.

Reviewer 3: Think we should add in a note for price ids that this might not apply for. Something along the lines of:

```suggestion
**Note**: An API for accessing data is not the same thing as the markets you choose to determine the price.  You would not say that cryptowatch is a market for data, as it is just an aggregator of market data.  Examples of markets are “Coinbase Pro, Binance, Bitstamp.”  Data sources are still necessary to demonstrate the accessibility of the data and are often just the specific market’s API.

For price identifiers that are not determined by market prices, the markets and data sources format may not apply. In these situations, this section should inform users of how they should gather all necessary information used in the price determination process.
```

<br>

# PRICE FEED IMPLEMENTATION

To allow for the creation of bots that can programmatically calculate prices off-chain to liquidate and dispute transactions, you must create a price feed following the UMA Protocol format (outlined below). This price feed is also necessary to calculate developer mining rewards.

To meet this criteria, please include a link to a PR to the UMA [protocol repo](https://github.com/UMAprotocol/protocol) with an example price feed that inherits this [PriceFeedInterface](https://github.com/UMAprotocol/protocol/blob/master/packages/financial-templates-lib/src/price-feed/PriceFeedInterface.js). 

Classes of price identifiers that are supported by default and require no additional price feed include.

- Any currency pair available on [Cryptowatch](https://github.com/UMAprotocol/protocol/blob/master/packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.js)
- [Uniswap prices](https://github.com/UMAprotocol/protocol/blob/master/packages/financial-templates-lib/src/price-feed/UniswapPriceFeed.js)
- [Balancer prices](https://github.com/UMAprotocol/protocol/blob/master/packages/financial-templates-lib/src/price-feed/BalancerPriceFeed.js)

> Please provide a link to your price feed pull request.

Reviewer 3: ```suggestion
Please follow the [UMA contribution guidelines](https://github.com/UMAprotocol/protocol/blob/master/CONTRIBUTING.md) when submitting your pull request, and provide a link in your UMIP.
```

3. **Input processing**
    - Denote if and why UMA tokenholders should change the calculation method for the price identifier. 
<br>

    - For example, for ETH/USD price requests, inputs are results received from exchanges and if an exchange is no longer valid, then human intervention is required. 

    - The input processing can typically be denoted as "Human intervention in extreme circumstances where the result differs from broad market consensus".
<br>

4. **Result processing**
    - What type of processing should voters perform to determine the final price of an asset (e.g., take the median or mode of all votes submitted)? 
    - **Note** - a result processing of "median" is more resilient to market manipulation versus a result processing of "average". 

<br>

# Security considerations

### **Example questions**
<br>

1. Where could manipulation occur?

2. How could this price ID be exploited?
3. Do the instructions for determining the price provide people with enough certainty?
4. What are current or future concern possibilities with the way the price identifier is defined?
5. Are there any concerns around if the price identifier implementation is deterministic?

<br>

> **Example Security Section**

Reviewer 3: ```suggestion
**Example**
```
